types/factory-girl/index.d.ts

```diff
@@ -143,7 +143,7 @@ declare namespace factory {
         afterCreate?: Hook<T> | undefined;
     }
 
-    type Hook<T> = (model: any, attrs: T | T[], options: any) => void;
+    type Hook<T> = (model: any, attrs: T | T[], options: any) => any;
```

Comment:
A generic like this should be valid. Any type can still be passed for `model`, but the return value must be the same type:

```suggestion
    type Hook<T> = <M>(model: M, attrs: T | T[], options: any) => M | Promise<M>;
```

Reply:
Nice I was looking for something like this 👍

Reply:
Unfortunately this does not allow modification of the model since we do not know at compile time what type M might be:
```
afterBuild: (model, attrs, options) => {
    // for buildMany & build
    model.email = Array.isArray(attrs) ? attrs[0].email : attrs.email;
    return model;
},
```
Gives us: `Property 'email' does not exist on type 'M'.ts(2339)`

And if I try to type model, then it fails because we are defining a type which might not match the actual given type:
```
afterBuild: (model: { email?: string }, attrs, options) => {
    // for buildMany & build
    model.email = Array.isArray(attrs) ? attrs[0].email : attrs.email;
    return model;
},
```
Gives: `Type '<M extends unknown>(model: { email?: string | undefined; }, attrs: User | User[], options: any) => { email?: string | undefined; }' is not assignable to type 'Hook<User>'.
  Types of parameters 'model' and 'model' are incompatible.
    Type 'M' is not assignable to type '{ email?: string | undefined; }'.ts(2322)`

The only way to solve this using generics is to make the entire Options interface take this generic and using it in the defines & extends methods. Do you think this is preferable?

Reply:
I'm not sure whether this would ever actually end up with a generic other than `any`, but it might be worth trying

Reply:
On second thought, this would probably just cause `M` to always equal `any` unless some other changes are made